x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/AutoConfigInitialNode.java

```diff
@@ -0,0 +1,335 @@
+            //throw new UserException(ExitCodes.OK, "Skipping security auto configuration: TLS already configured");
+            return;
+        }
+        // check that no file realms have been configured.
+        // this can technically be improved upon:
+        // auto configuration only requires that the file realm be enabled, and, for optimal experience,
+        // also be the first in the chain
+        if (false == env.settings().getByPrefix(RealmSettings.realmSettingPrefix(FileRealmSettings.TYPE)).isEmpty()) {
+            terminal.println(Terminal.Verbosity.VERBOSE,
+                    "Skipping security auto configuration because it appears that a file-based realm is already configured.");
+            //throw new UserException(ExitCodes.OK, "Skipping security auto configuration: file realm configured");
+            return;
+        }
+        // tolerate enabling enrollment explicitly, as it could be useful to enable it by a command line option
+        // only the first time that the node is started
+        if (env.settings().hasValue(XPackSettings.ENROLLMENT_ENABLED.getKey()) && false ==
+                XPackSettings.ENROLLMENT_ENABLED.get(env.settings())) {
+            terminal.println(Terminal.Verbosity.VERBOSE,
+                    "Skipping security auto configuration because enrollment is explicitly disabled.");
+            //throw new UserException(ExitCodes.OK, "Skipping security auto configuration: enrollment disabled");
+            return;
+        }
+
+        final ZonedDateTime autoConfigDate = ZonedDateTime.now();
+        final String instantAutoConfigName = "auto_generated_" + autoConfigDate.toInstant().getEpochSecond();
+        final Path instantAutoConfigDir = env.configFile().resolve(instantAutoConfigName);
+        try {
+            Files.createDirectory(instantAutoConfigDir);
+        } catch (IOException e) {
+            throw new UserException(ExitCodes.CANT_CREATE, "Could not create auto configuration directory", e);
+        }
+
+        // the transport key-pair is the same across the cluster and is trusted without hostname verification (it is self-signed),
+        // do not populate the certificate's IP, DN, and CN certificate fields
+        final X500Principal certificatePrincipal = new X500Principal("CN=" + System.getenv("HOSTNAME"));
+        final GeneralNames subjectAltNames = getSubjectAltNames();
+
+        KeyPair transportKeyPair = CertGenUtils.generateKeyPair(TRANSPORT_KEY_SIZE);
+        // self-signed which is not a CA
+        X509Certificate transportCert = CertGenUtils.generateSignedCertificate(certificatePrincipal,
+                subjectAltNames, transportKeyPair, null, null, false, TRANSPORT_CERTIFICATE_DAYS, null);
+        KeyPair httpCAKeyPair = CertGenUtils.generateKeyPair(HTTP_CA_KEY_SIZE);
+        // self-signed CA
+        X509Certificate httpCACert = CertGenUtils.generateSignedCertificate(certificatePrincipal,
```

Comment:
I thought about having a particular way to identify CAs that have been auto-generated by this.
I tried using a SAN for it. It doesn't look great, I see what I can do.

Reply:
I've used `CertGenUtils.createCommonName(AutoConfigInitialNode.class.getName())` which doesn't print well because it is non-standard .
It should suffice for a check in the code though.